api/src/websocket/controllers/graphql.ts

```diff
@@ -0,0 +1,89 @@
+import SocketController from './base';
+import type { AuthenticationState, AuthMode, ConnectionParams, UpgradeContext, WebSocketClient } from '../types';
+import { handleWebsocketException, WebSocketException } from '../exceptions';
+import { authenticateConnection, refreshAccountability } from '../authenticate';
+import { waitForAnyMessage } from '../utils/wait-for-message';
+import { BasicAuthMessage, WebSocketMessage } from '../messages';
+
+export class GraphQLSubscriptionController extends SocketController {
+	gql: Server<ConnectionParams>;
+	constructor(httpServer: httpServer) {
+		super(httpServer, 'WS GraphQL', String(env.WEBSOCKETS_GRAPHQL_PATH), {
+			mode: String(env.WEBSOCKETS_GRAPHQL_AUTH).toLowerCase() as AuthMode,
+			timeout: Number(env.WEBSOCKETS_GRAPHQL_AUTH_TIMEOUT) * 1000,
+			verbose: false,
+		});
+		if ('WEBSOCKETS_GRAPHQL_CONN_LIMIT' in env) {
+			this.maxConnections = Number(env.WEBSOCKETS_GRAPHQL_CONN_LIMIT);
+		}
+		this.server.on('connection', (ws: WebSocket, auth: AuthenticationState) => {
+			this.bindEvents(this.createClient(ws, auth));
+		});
+		this.gql = makeServer<ConnectionParams>({
+			schema: async (ctx) => {
+				const accountability = await getAccountabilityForToken(ctx.connectionParams?.access_token);
+
+				// for now only the items will be watched in the MVP system events tbd
+				const service = new GraphQLService({
+					schema: await getSchema(),
+					scope: 'items',
+					accountability,
+				});
+
+				return service.getSchema();
+			},
+		});
+		logger.info(`GraphQL Subscriptions started at ws://${env.HOST}:${env.PORT}${this.endpoint}`);
+	}
+	private bindEvents(client: WebSocketClient) {
+		const closedHandler = this.gql.opened(
+			{
+				protocol: client.protocol, // will be validated
+				send: (data) =>
+					new Promise((resolve, reject) => {
+						client.send(data, (err) => (err ? reject(err) : resolve()));
+					}),
+				close: (code, reason) => client.close(code, reason), // for standard closures
+				onMessage: (cb) => {
+					client.on('parsed-message', async (message: WebSocketMessage) => {
+						try {
+							client.accountability = await refreshAccountability(client.accountability);
+							await cb(JSON.stringify(message));
+						} catch (error) {
+							handleWebsocketException(client, error, 'graphql');
+						}
+					});
+				},
+			},
+			{}
+		);
+
+		// notify server that the socket closed
+		client.once('close', (code, reason) => closedHandler(code, reason.toString()));
+	}
+	protected override async handleHandshakeUpgrade({ request, socket, head }: UpgradeContext) {
+		this.server.handleUpgrade(request, socket, head, async (ws) => {
+			try {
+				const msg = await waitForAnyMessage(ws, this.authentication.timeout);
+				if (msg?.type !== 'connection_init') throw new Error();
+
+				const state = await authenticateConnection(BasicAuthMessage.parse(msg.payload));
+				this.server.emit('connection', ws, state);
+				ws.send(JSON.stringify({ type: 'connection_ack' }));
```

Comment:
I that would work however the intention of that `fmtMessage` function is to add extra data in an organized way. I did not use the function here because GraphQL does not play well with extra data in this response.

Reply:
Thanks for the clarification! I was indeed curious whether it's intentional in GraphQL controller since it's the only one, so I was hesitant on the need for it. Makes sense, marking this as resolved 👍